### PR TITLE
Fixes to build using GCC 8

### PR DESCRIPTION
**Pull request by tautschnig**
GCC 8 has a few additional warnings. This PR addresses genuine shortcomings and silences the spurious warnings.

### CODE REVIEW

File: src/java_bytecode/expr2java.h [hunk 1]
```diff
@@ -58,7 +58,8 @@ std::string type2java(const typet &type, const namespacet &ns);
 template <typename float_type>
 std::string floating_point_to_java_string(float_type value)
 {
-  const auto is_float = std::is_same<float_type, float>::value;
+  static const auto is_float = std::is_same<float_type, float>::value;
+  (void)is_float; // silence a spurious warning generated by GCC 8
   static const std::string class_name = is_float ? "Float" : "Double";
```
Comment (kroening) [inline src/java_bytecode/expr2java.h]: Uh, this one is sad

**Review by kroening — APPROVED** [review]

File: src/java_bytecode/expr2java.h (hunk 1 above)
Comment (tautschnig) [reply src/java_bytecode/expr2java.h]: Yes. I changed the name of the variables where it is being used and that does cause a compile failure. So it really seems to be a bug...

File: scripts/minisat-2.2.1-patch [hunk 2]
```diff
@@ -227,3 +227,19 @@ index 9cbbc51..27b9700 100644
...
+index 620e0c8021..566a807dac 100644
+--- a/minisat/mtl/Vec.h
++++ b/minisat/mtl/Vec.h
+@@ -96,8 +96,11 @@ template<class T>
+ void vec<T>::capacity(int min_cap) {
+     if (cap >= min_cap) return;
+     int add = imax((min_cap - cap + 1) & ~1, ((cap >> 1) + 2) & ~1);   // NOTE: grow by approximately 3/2
++#include <util/pragma_push.def>
++#include <util/pragma_wclass_memaccess.def>
+     if (add > INT_MAX - cap || (((data = (T*)::realloc(data, (cap += add) * sizeof(T))) == NULL) && errno == ENOMEM))
```
Comment (smowton) [inline scripts/minisat-2.2.1-patch]: http://pubs.opengroup.org/onlinepubs/009695299/functions/realloc.html indicates `errno` may only have value `ENOMEM` when `realloc` fails, so the test `... && errno == ENOMEM` here seems to just open a door for non-conformant implementations to leave us with a null `data` but no OOM exception

File: src/java_bytecode/expr2java.h [hunk 3]
```diff
@@ -58,7 +58,8 @@ std::string type2java(const typet &type, const namespacet &ns);
 template <typename float_type>
 std::string floating_point_to_java_string(float_type value)
 {
-  const auto is_float = std::is_same<float_type, float>::value;
+  static const auto is_float = std::is_same<float_type, float>::value;
```
Comment (smowton) [inline src/java_bytecode/expr2java.h]: Brief experimentation shows the failure also goes away if you replace `const auto is_float` with `[static] const bool is_float`, so I agree it's a bug but suggest that's a cleaner fix.

File: scripts/minisat-2.2.1-patch [hunk 4]
```diff
@@ -227,3 +227,19 @@ index 9cbbc51..27b9700 100644
...
+diff --git a/minisat/mtl/Vec.h b/minisat/mtl/Vec.h
+index 620e0c8021..566a807dac 100644
+--- a/minisat/mtl/Vec.h
++++ b/minisat/mtl/Vec.h
+@@ -96,8 +96,11 @@ template<class T>
+ void vec<T>::capacity(int min_cap) {
+     if (cap >= min_cap) return;
+     int add = imax((min_cap - cap + 1) & ~1, ((cap >> 1) + 2) & ~1);   // NOTE: grow by approximately 3/2
++#include <util/pragma_push.def>
++#include <util/pragma_wclass_memaccess.def>
```
Comment (smowton) [inline scripts/minisat-2.2.1-patch]: Better to just cast the `data` pointer to `char *`? I looked at this in a bit more detail; concluded that the fundamental problem is `vec<T>`'s non-trivial destructor -- the type is not trivially copyable, but is trivially movable, but regrettably C++ doesn't have a concept for that (yet). However the cast, at least for my system (Ubuntu 18.04 / `gcc-8 (Ubuntu 8-20180414-1ubuntu2) 8.0.1 20180414 (experimental) [trunk revision 259383]`) the cast was enough to circumvent the error.

**Comment by tautschnig** [comment]
@smowton Thank you very much, I have applied all your suggestions.

**Comment by smowton** [comment]
You didn't apply the ENOMEM one?

File: jbmc/src/java_bytecode/expr2java.h [hunk 5]
```diff
@@ -83,7 +83,7 @@ std::string floating_point_to_java_string(float_type value)
     {
       return std::stod(decimal) == value;
     }
-    catch(std::out_of_range)
+    catch(std::out_of_range&)
```
Comment (smowton) [inline jbmc/src/java_bytecode/expr2java.h]: Non-blocking nitpick: space before `&`

**Comment by tautschnig** [comment]
> You didn't apply the ENOMEM one?

Ah, no, because that seemed unrelated to the PR subject? Let me get to this right away in a fresh PR.

File: jbmc/src/java_bytecode/expr2java.h (hunk 5 above)
Comment (tautschnig) [reply jbmc/src/java_bytecode/expr2java.h]: done

**Review by smowton — APPROVED** [review]

**Comment by tautschnig** [comment]
> Ah, no, because that seemed unrelated to the PR subject? Let me get to this right away in a fresh PR.

Actually backtracking on this: I've made Minisat use its own xrealloc (which for some mysterious reason wasn't being used here), and I've fixed that implementation as well.

File: scripts/minisat-2.2.1-patch [hunk 6]
```diff
@@ -227,3 +227,33 @@ index 9cbbc51..27b9700 100644
...
+--- a/minisat/mtl/XAlloc.h
++++ b/minisat/mtl/XAlloc.h
+@@ -32,8 +32,9 @@
+ class OutOfMemoryException{};
+ static inline void* xrealloc(void *ptr, size_t size)
+ {
++    errno = 0;
+     void* mem = realloc(ptr, size);
+-    if (mem == NULL && errno == ENOMEM){
++    if (mem == NULL || errno == ENOMEM){
```
Comment (smowton) [inline scripts/minisat-2.2.1-patch]: Just drop the `errno` test completely -- it may have any value when a call succeeds, not necessarily `0`.

Comment (tautschnig) [reply scripts/minisat-2.2.1-patch]: Done.

**Review by smowton — APPROVED** [review]
Looks good -- re-check that the void-pointer cast implied by xrealloc does the trick though, compared to the char-pointer cast introduced before

**Comment by tautschnig** [comment]
> re-check that the void-pointer cast implied by xrealloc does the trick though

At least for me that seems to be the case (g++-8 (Ubuntu 8-20180424-0ubuntu1~14.04) 8.0.1 20180424)

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**